### PR TITLE
Implement calculations over arbitrary sub-image areas / frame sizes

### PR DESCRIPTION
**Pull request by simonrw**
Issue #35

### CODE REVIEW

**Comment by mindriot101** [comment]
This is just a start. I've allowed the imaging region to be specified using image extent, but haven't implemented the tile size logic yet. I think it'll take some thinking about.

**Comment by jmccormac01** [comment]
Nice! I think the idea was to have a supplied minimum tile size, then work out how many you need, making the tile size slightly bigger if needed. The tile size should be adjusted to minimise the wastage chopped of from the upper Y and upper X edges. 

For example, we have a 700 pixel square with a tile size of 32 pixels. 

700/32  = 21.875    -    32*21 = 672 (28 pixels chopped off)
700/33  = 21.212    -    33*21 = 693 (7 pixels chopped off)
700/34 = 20.588     -    20*34 = 680 (20 pixels chopped off)

So in this case the best tile size would be 33 pixels.  We could automatically scan through +10% of the provided pixel size. This additional adjustment is not super critical. The person using the code should know the CCD size and tile size they want if using this relatively special feature. 

It will be interesting to see how this performs in the infinite number of combinations of settings possible with this mode. I did test the single guide star case for the paper and we need a SNR in a small postage stamp containing only 1 star of ~10 to work and I quoted SNR_limiting = 15 to be safe.

**Comment by mindriot101** [comment]
Discussion moved to PR #44

File: donuts/image.py [hunk 1]
```diff
@@ -91,39 +97,70 @@ def trim(self, prescan_width=0, overscan_width=0, border=64):
...
+            nparams = len(region_extent)
+            if nparams != 4:
+                raise TypeError(
+                    'Incorrect number of parameters for `region_extent`. '
+                    'Should be 4, got {}'.format(nparams)
+                )
+
+            x1, x2, y1, y2 = region_extent
+
+            if x2 < x1 or y2 < y1:
```
Comment (pchote) [inline donuts/image.py]: It would be a good idea to check against the image dimensions here too.

Comment (mindriot101) [inline donuts/image.py]: Good point, done: 5a4a63035486b1f82891708bce9764bf496f5dd0

File: donuts/image.py [hunk 2]
```diff
@@ -91,39 +97,75 @@ def trim(self, prescan_width=0, overscan_width=0, border=64):
...
+                raise TypeError(
+                    'Invalid region dimensions. x1 should be < x2 '
+                    'and y1 should be < y2'
+                )
+
+            if (x1 < 0) or (y1 < 0) or (x2 >= self.raw_image.shape[1]) or (y2 >= self.raw_image.shape[0]):
```
Comment (pchote) [inline donuts/image.py]: The `>=` flags full-frame regions as invalid, so these need to be `>`.

File: donuts/tests/test_image.py [hunk 3]
```diff
@@ -75,6 +75,24 @@ def test_returns_self(self):
         testvalue = image.trim()
         assert testvalue is image
 
+    def test_manual_region_choosing(self):
```
Comment (pchote) [inline donuts/tests/test_image.py]: It would be useful to have some tests on non-square images to check for array indexing issues.

**Comment by pchote** [comment]
If the region isn't an integer multiple of the tile size this blows up at
```python
self.backsub_region = self.raw_region - self.sky_background
```

File: donuts/image.py (hunk 2 above)
Comment (mindriot101) [inline donuts/image.py]: Did we have a decision on whether to use x/y + width/height?

Comment (pchote) [inline donuts/image.py]: I think we all agreed [x,y,w,h] would be a good idea, yeah.

**Comment by pchote** [comment]
> but haven't implemented the tile size logic yet. I think it'll take some thinking about.

The easy way out would be to require the region to be an exact multiple of the tile size, and leave it to the user to sort out the details.

**Comment by jmccormac01** [comment]
I'm going to redo this today. Closing this PR

**Pull request closed without merging**